### PR TITLE
feat(tls) add duplex upgrade

### PR DESCRIPTION
**Pull request by cirospaciari**
### What does this PR do?
Fix: https://github.com/oven-sh/bun/issues/5147
Fix: https://github.com/oven-sh/bun/issues/13736
Fix: https://github.com/oven-sh/bun/issues/4033

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [X] Code changes

### How did you verify your code works?
- [X] - node:tls connect tests
- [ ] - MSSQL integrated tests
- [ ] - MySQL integrated tests

### CODE REVIEW

**Comment by robobun** [comment]
@cirospaciari, your commit 636f459aae3e3ed0bdea90d1fcdcb83fccbea0f4 is building: [`#2697`](https://buildkite.com/bun/bun/builds/2697)

File: src/deps/uws.zig [hunk 1]
```diff
@@ -69,21 +73,455 @@ pub const InternalLoopData = extern struct {
...
+            .onOpen = UpgradedDuplex.onOpen,
+            .onHandshake = UpgradedDuplex.onHandshake,
+            .onData = UpgradedDuplex.onData,
+            .onClose = UpgradedDuplex.onClose,
+            .write = UpgradedDuplex.internalWrite,
+        });
+        if (this.origin.get()) |duplex| {
+            const globalThis = this.origin.globalThis.?;
+
+            const addEventListener = try duplex.getFunction(globalThis, "on") orelse return error.InvalidDuplex;
```
Comment (Jarred-Sumner) [inline src/deps/uws.zig]: Can we do this step in JavaScript instead? Calling functions like this and using runCallback here will not handle exceptions properly

File: src/deps/uws.zig [hunk 2]
```diff
@@ -69,21 +73,455 @@ pub const InternalLoopData = extern struct {
...
+            this.onWritableCallback.deinit();
+        }
+        if (this.onCloseCallback.get()) |callback| {
+            JSC.setFunctionData(callback, null);
+            this.onCloseCallback.deinit();
+        }
+        const ssl_error = this.ssl_error;
+        this.ssl_error = .{};
+        if (ssl_error.reason.len > 0) {
+            bun.default_allocator.free(ssl_error.reason);
```
Comment (Jarred-Sumner) [inline src/deps/uws.zig]: can we move this code to a .deinit function on the ssl_error struct?

Comment (cirospaciari) [reply src/deps/uws.zig]: done

File: src/deps/uws.zig (hunk 1 above)
Comment (cirospaciari) [reply src/deps/uws.zig]: done

**Pull request merged**